DR-1475: Billing Profile access checks for verifyAccount

fboulnois opened the pull request:
Implement access checks on ProfileService.verifyAccount

fboulnois committed:
Add billing profile access checks

fboulnois committed:
Switch to ImpersonatedCredentials

fboulnois committed:
Add comment to explain functionality

fboulnois committed:
Add a bunch of logging

pshapiro4broad commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -14,38 +16,54 @@
+                    credentials = credentials.createScoped(scopes);
+                }
+            } else {
+                Date soon = Date.from(new Date().toInstant().plusSeconds(300));
```
slightly simpler 
```suggestion
                Date soon = Date.from(Instant.now().plusSeconds(300));
```

pshapiro4broad commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -14,38 +16,54 @@
-            if (credentials.createScopedRequired()) {
-                credentials = credentials.createScoped(scopes);
+
+            GoogleCredentials credentials = null;
```
This is always assigned so it shouldn't need an initial value here. If you slightly rewrite the code below you can make this final, which would make it apparent to the reader that this is definitely assigned.
```
if (user == null) {
    if (serviceAccountCredentials.createScopesRequired()) {
        credentials = serviceAccountCredentials.createScoped(scopes);
    } else {
        credentials = serviceAccountCredentials;
    }
} else { ...
```

pshapiro4broad commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -54,6 +72,32 @@ private static CloudBillingClient cloudBillingClient() {
+        List<String> permissions = Collections.singletonList("billing.resourceAssociations.create");
+        TestIamPermissionsRequest permissionsRequest = TestIamPermissionsRequest.newBuilder()
+            .setResource(resource.toString())
+            .addAllPermissions(permissions)
```
Since it's one permission you can write
```suggestion
            .addPermissions("billing.resourceAssociations.create")
```

fboulnois replied:
That adds a bit of complexity for checking the response as it returns a list of permissions.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -54,6 +72,32 @@ private static CloudBillingClient cloudBillingClient() {
+
+    public static boolean verifyAccess(AuthenticatedUserRequest user, String billingAccountId) {
+        ResourceName resource = BillingAccountName.of(billingAccountId);
+        List<String> permissions = Collections.singletonList("billing.resourceAssociations.create");
```
It would be great to have a constant for magic strings like this. It's too bad Google hasn't provided them

pshapiro4broad reviewed (commented):
Minor java related comments

fboulnois committed:
Address review comments

fboulnois commented:
> Minor java related comments

Thank you for the comments! I rebased on develop and included your changes.

fboulnois committed:
Add additional logging

fboulnois committed:
Switch cloud-platform to use userLoginScopes

fboulnois committed:
Simplify logging

fboulnois committed:
Switch to ImpersonatedCredentials

fboulnois committed:
Ensure scopes are always created on service account

fboulnois committed:
Move comments and rename methods

fboulnois committed:
Move scopes to serviceAccountCredentials

fboulnois committed:
Add cloud-platform scope to swagger ui

fboulnois commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -71,18 +103,20 @@ private static CloudBillingClient cloudBillingClient() {
         try {
-            TestIamPermissionsResponse response = cloudBillingClient().testIamPermissions(permissionsRequest);
+            logger.info("Testing IAM permission on billing account: {}", billingAccountId);
+            TestIamPermissionsResponse response = cloudBillingClient(user).testIamPermissions(permissionsRequest);
```
The test that I am trying to execute is `IT_JADEAPIURL=http://localhost:8080 ./gradlew testIntegration --tests *FileTest.fileParallelFailedLoadTest --scan` . It fails here.

nmalfroy committed:
Don't try to impersonate the user but expect a properly scoped token

It would be great if we could verify that the token should be  set but this at least seems to work

nmalfroy commented:
Just put up a commit that seems to work.  The idea is to not use impersonation and use a scoped token, which is I think what you originally had.  In talking to Doug, this is what Rawls does

fboulnois committed:
Remove extra logging

fboulnois committed:
Ensure that client is cleaned up automatically

fboulnois committed:
Remove verifyAccount TODO comment

fboulnois committed:
Remove authorizeLinking TODO comment

fboulnois committed:
Also add scope to client tests

fboulnois committed:
Formatting for spotless checks

fboulnois committed:
Replace billing account exception text

nmalfroy reviewed (approved):
lgtm!

ddietterich commented on datarepo-clienttests/src/main/java/common/utils/AuthenticationUtils.java:
```diff
@@ -28,7 +28,8 @@ private AuthenticationUtils() {}
   // what the UI requests, so our tests represent actual user behavior
-  private static final List<String> userLoginScopes = Arrays.asList("openid", "email", "profile");
+  private static final List<String> userLoginScopes =
+      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-platform");
```
I'm not sure what to do about it, but we have this list in at least 4 places.
I wonder if there is a way to DRY that out.

fboulnois replied:
Yes! I was thinking something very similar. I'll make a separate ticket.

fboulnois replied:
I created DR-1548 to track this.

ddietterich commented on src/main/resources/application.properties:
```diff
@@ -26,7 +26,7 @@ db.stairway.forceClean=true
 oauth.schemeName=googleoauth
 oauth.loginEndpoint=https://accounts.google.com/o/oauth2/auth
-oauth.scopes=openid,email,profile
+oauth.scopes=openid,email,profile,https://www.googleapis.com/auth/cloud-platform
```
I wonder why this is in `application.properties`. Probably just historical.
These values are required by the code. It is not like we change this across different deployments.
This config is only used to feed SwaggerDocumentationConfig which looks like generated code that is no longer generated. Hmm...

fboulnois replied:
As I recall, swagger-ui actually needs this to have this scope appear for dev (non-test) users.

ddietterich reviewed (approved):
lgtm

One comment that maybe we make a separate ticket for. Do we need to repeat the scope list four times?
